File: go/1266.go
package main

func minTimeToVisitAllPoints(points [][]int) int {
	result := 0

	x := points[0][0]
	y := points[0][1]

	for i := 1; i < len(points); i++ {
		for x != points[i][0] || y != points[i][1] {
			result++
			// horizontal
			if x == points[i][0] {
				if points[i][1] < y {
					y--
				} else {
					y++
				}
				// vertical
			} else if y == points[i][1] {
				if points[i][0] < x {
					x--
				} else {
					x++
				}
				// diagonal
			} else {
				if x < points[i][0] && y < points[i][1] {
					x++
					y++
				} else if x > points[i][0] && y > points[i][1] {
					x--
					y--
				} else if x < points[i][0] && y > points[i][1] {
					x++
					y--
				} else {
					x--
					y++
				}
			}
		}
	}

	return result
}

File: zig/1266.zig
const std = @import("std");

fn chebyshevDistance(a: [2]i32, b: [2]i32) u32 {
    return @max(@abs(a[0] - b[0]), @abs(a[1] - b[1]));
}

fn solution(points: [][2]i32) u32 {
    var result: u32 = 0;
    var p: [2]i32 = points[0];

    for (1..points.len) |i| {
        result += chebyshevDistance(p, points[i]);
        p = points[i];
    }

    return result;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    var points = [_][2]i32{
        .{ 1, 1 },
        .{ 3, 4 },
        .{ -1, 0 },
    };

    const expected = 7;
    const actual = solution(&points);

    try expect(expected, actual);
}

test "test case #2" {
    var points = [_][2]i32{
        .{ 3, 2 },
        .{ -2, 2 },
    };

    const expected = 5;
    const actual = solution(&points);

    try expect(expected, actual);
}
